File: src/signal_simple/signal_simple_workflow.erl
-module(signal_simple_workflow).

% elp:ignore W0012 W0040
-moduledoc false.

-export([
    execute/2
]).

-include_lib("temporal_sdk/include/workflow.hrl").

-define(CANCEL_ACTIVITY_SIGNAL, "cancel_activity").

execute(_Context, _Input) ->
    A = start_activity(echo_activity, [[], 3_000], [{heartbeat_timeout, 1_000}]),
    case await_one([{signal_request, ?CANCEL_ACTIVITY_SIGNAL}, A], {10, second}) of
        {ok, [#{}, #{state := S}]} when S =:= cmd; S =:= scheduled; S =:= started ->
            cancel_activity(A),
            set_workflow_result(["Cancel requested. Activity canceled."]);
        {ok, [noevent, #{}]} ->
            set_workflow_result(["Cancel not requested. Activity closed."]);
        %% For the sake of educational completeness:
        {ok, [#{}, noevent]} ->
            set_workflow_result(["Cancel requested. Activity not started."]);
        {noevent, _} ->
            set_workflow_result(["Cancel not requested. Activity not started."])
    end.

File: lib/signal_simple/signal_simple_workflow.ex
defmodule SignalSimple.Workflow do
  use TemporalSdk.Workflow
  @moduledoc false

  @cancel_activity_signal "cancel_activity"

  @impl true
  def execute(_context, _input) do
    a = start_activity(:echo_activity, [[], 3_000], heartbeat_timeout: 1_000)

    case await_any([{:signal_request, @cancel_activity_signal}, a], {10, :second}) do
      {:ok, [%{}, %{state: s}]} when s == :cmd or s == :scheduled or s == :started ->
        cancel_activity(a)
        set_workflow_result(["Cancel requested. Activity canceled."])

      {:ok, [:noevent, %{}]} ->
        set_workflow_result(["Cancel not requested. Activity closed."])

      # For the sake of educational completeness:
      {:ok, [%{}, :noevent]} ->
        set_workflow_result(["Cancel requested. Activity not started."])

      {:noevent, _} ->
        set_workflow_result(["Cancel not requested. Activity not started."])
    end
  end
end
